crypto/keys/src/traits.rs

```diff
@@ -48,60 +80,88 @@ pub trait DistinguishedEncoding: Sized {
     /// Create a new object from the given DER-encoded SubjectPublicKeyInfo.
     fn try_from_der(src: &[u8]) -> Result<Self, KeyError>;
 
+    /// Create the standardized DER-encoded representation of this object and
+    /// write to the provided buffer.
+    fn to_der_slice<'a>(&self, buff: &'a mut [u8]) -> &'a [u8];
+
     /// Create the standardized DER-encoded representation of this object.
-    fn to_der(&self) -> Vec<u8>;
+    #[cfg(feature = "alloc")]
+    fn to_der(&self) -> Vec<u8> {
+        let mut buff = alloc::vec![0; <Self as DistinguishedEncoding>::der_size()];
+        self.to_der_slice(&mut buff);
+        buff
+    }
 }
 
-/// A trait indicating that a string fingerprint can be generated for an
-/// object.
+/// Maximum length of the DER buffer required for
+/// [`DistinguishedEncoding::to_der()`]
+pub const DER_MAX_LEN: usize = 128;
+
+/// A trait indicating that a fingerprint can be generated for an object.
 pub trait Fingerprintable {
-    fn fingerprint<D: Digest>(&self) -> Result<String, KeyError>;
+    /// Generate fingerprint
+    fn fingerprint<D: Digest>(&self) -> Fingerprint<D>;
+}
+
+/// A fingerprint object, generic over digest output size
+pub struct Fingerprint<D: digest::OutputSizeUser> {
+    hash: digest::generic_array::GenericArray<u8, D::OutputSize>,
+}
+
+/// Debug impl for fingerprint objects
+impl<D: digest::OutputSizeUser> core::fmt::Debug for Fingerprint<D> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        for i in 0..self.hash.len() {
+            write!(f, "{:02x}", self.hash[i])?;
+            if i < self.hash.len() - 1 {
+                write!(f, ":")?;
+            }
+        }
+        Ok(())
```

Comment:
Optional/Equivalent: Forward to the `Display` impl. There's also a convention of including the type in `Debug`
```suggestion
        write!(f, "Fingerprint({})", self)
```